docs/language/ql-training-rst/cpp/bad-overflow-guard.rst

```diff
@@ -0,0 +1,228 @@
+
+
+Checking for overflow in C
+==========================
+
+- Arithmetic operations overflow if the result is too large for their type.
+- Developers sometimes exploit this to write overflow checks:
+
+.. code-block:: cpp
+
+  if (v + b < v) {
+      handle_error("overflow");
+  } else {
+      result = v + b;
+  }
+
+Where might this go wrong?
+
+.. note::
+
+  - In C/C++ we often need to check for whether an operation `overflows <https://en.wikipedia.org/wiki/Integer_overflow>`__.
+  - An overflow is when an arithmetic operation, such as an addition, results in a number which is too large to be stored in the type.
+  - When an operation overflows, the value “wraps” around.
+  - A typical way to check for overflow of an addition, therefore, is whether the result is less than one of the arguments - that is the result has **wrapped**.
+
+Integer promotion
+=================
+
+From `https://en.cppreference.com/w/c/language/conversion <https://en.cppreference.com/w/c/language/conversion>`__:
+
+*Integer promotion is the implicit conversion of a value of any integer type with rank less or equal to rank of* ``int`` *... to the value of type* ``int`` *or* ``unsigned int``.
+
+The arguments of the following arithmetic operators undergo implicit conversions:
+
+- binary arithmetic (* / % + - )
+- relational operators (< > <= >= == !=)
+- binary bitwise operators (& ^ \|)
+- the conditional operator (?:)
+
+.. note::
+
+  - Most of the time integer conversion works fine. However, the rules governing addition in C/C++ are complex, and it easy to get bitten.
+  - CPUs generally prefer to perform arithmetic operations on 32 bit or larger integers, as the architectures are optimised to perform those efficiently.
+  - The compiler therefore performs “integer promotion” for arguments to arithmetic operations that are smaller than 32 bit.
+
+Checking for overflow in C revisited
+====================================
+
+Which branch gets executed in this example? What is the value of ``result``?
+
+.. code-block:: cpp
+
+  uint16_t v = 65535;
+  uint16_t b = 1;
+  uint16_t result;
+  if (v + b < v) {
+      handle_error("overflow");
+  } else {
+      result = v + b;
+  }
+
+.. note::
+
+  In this example the second branch is executed, even though there is a 16-bit overflow, and ``result`` is set to zero.
+
+Checking for overflow in C revisited
+====================================
+
+Here is the example again, with the conversions made explicit:
+
+.. code-block:: cpp
+
+  uint16_t v = 65535;
+  uint16_t b = 1;
+  uint16_t result;
+  if ((int)v + (int)b < (int)v) {
+      handle_error("overflow");
+  } else {
+      result = (uint16_t)((int)v + (int)b);
+  }
+
+.. note::
+
+  In this example the second branch is executed, even though there is a 16-bit overflow, and result is set to zero.
+
+  Explanation:
+  - The two integer arguments to the addition, v and b, are promoted to 32 bit integers.
```

Comment:
Suggest using monospaced font for code elements in the speaker notes - it makes them easier to read.